### ISSUE TITLE
panic: range start index 255 out of range for slice of length 124

### ISSUE THREAD

**Issue by kornelski** [image-rs/image-gif#106]
in gif-0.11.2/src/reader/decoder.rs:251:28 the `bytes` offset may be outside of the buffer.

The file that exhibits this error [is NSFW](https://www.customfighters.com//media/flashing-squeezing-and-jiggling-gif-gif.56201/full).

**Comment by fintelia** [comment]
The panic is on this line:

**Comment by okaneco** [comment]
There seem to be 2 main points:

1. `StreamingDecoder::next_state` seems to be returning the wrong byte offset in the `DecodeSubBlock` match
2. How to handle unknown blocks in `next_state`'s `BlockStart` match arm

### `bytes` offset out of range

The panicking line of code is within a loop that calls `gif::reader::decoder::StreamingDecoder::next_state`. In that function is the following snippet:

https://github.com/image-rs/image-gif/blob/033353951d73be44d532720bd00c80617d04da17/src/reader/decoder.rs#L581-L588

I think on **L587** instead of `left`, it should be returning `n` as the bytes offset.

```diff
- return goto!(left, DecodeSubBlock(0), emit Decoded::Data(&[]));
+ return goto!(n, DecodeSubBlock(0), emit Decoded::Data(&[]));
```

With this change,
- both images in https://github.com/image-rs/image/issues/1551 return `Error: Format(DecodingFormatError { underlying: "unexpected EOF" })` instead of panicking,
- I'm able to decode and encode the gif from https://github.com/image-rs/image/issues/1452.

I haven't looked into other reported issues.

### Handling unknown blocks

After that patch, this file will decode the first frame but then return `Error: Format(DecodingFormatError { underlying: "unknown block type encountered" })`,  which brings us to how unknown block types are handled. 

https://github.com/image-rs/image-gif/blob/033353951d73be44d532720bd00c80617d04da17/src/reader/decoder.rs#L505-L509

I don't know what the proper block to use here should be or if more logic is required.

```rust
if self.allow_unknown_blocks {
    goto!(SkipBlock(b as usize))
} else {
    Err(DecodingError::format("unknown block type encountered"))
}
```

- imagemagick `identify` shows 75 frames in the file
- I'm only able to get 74 decoded/encoded with the mentioned changes
- Using imagemagick `convert input.gif -coalesce image%02d.jpg` to explode the gif, and the `explode` example with `allow_unknown_blocks`:
  - the 70th frame gets dropped, but I don't know the cause

---

The testbed I'm using is based off the `explode.rs` example in this repo and a draft PR.
<details><summary>
<i>Testbed program</i>
</summary>

```rust
fn main() -> Result<(), Box<dyn std::error::Error>> {
    let input_path = std::path::PathBuf::from(
        std::env::args_os()
            .nth(1)
            .ok_or("Specify a GIF path as the first argument")?,
    );
    let output_path = std::path::PathBuf::from(
        std::env::args_os()
            .nth(2)
            .ok_or("Specify a GIF path as the second argument")?,
    );

    let input = std::fs::File::open(&input_path)?;
    let mut options = gif::DecodeOptions::new();
    options.allow_unknown_blocks(false); // work in progress PR
    let mut decoder = options.read_info(input)?;
    let width = decoder.width();
    let height = decoder.height();
    let global_pal = decoder.global_palette().unwrap_or_default().to_vec();

    let mut decoded_number = 0;
    let mut encoded_number = 0;

    let mut file_out = std::fs::File::create(output_path)?;
    let mut encoder = gif::Encoder::new(&mut file_out, width, height, &global_pal)?;
    encoder.set_repeat(gif::Repeat::Infinite)?;

    while let Some(frame) = decoder.read_next_frame()? {
        decoded_number += 1;
        dbg!(decoded_number);

        encoder.write_frame(&frame)?;
        encoded_number += 1;
    }

    eprintln!("encoded frames: {}", encoded_number);

    Ok(())
}
```
</details>

**Comment by HeroicKatora** [comment]
Was fixed by #112, release soon.
